spa_client_concerns.md

```diff
@@ -0,0 +1,29 @@
+
+To prevent this, you need to register your app as a "public client". This means it has no client secret, and instead relies on other means to reliably identify the legitimate client. Specifically, it relies on a specific, registered "redirect_uri" that points to your SPA. For a mere redirect uri to be sufficient, it should also only use the HTTPS protocol; otherwise, I could set up a man-in-the-middle system and attack your users by intercepting their traffic (stealing their access tokens in the process).
+
+One downside of being a public client is that it has fewer options available for getting access tokens. Essentially, the choice is between the "Implicit Grant" and "Authorization Code Grant". These are the only two that are capable of relying on redirect_uris instead of client secrets.
+
+If you read the OAuth 2 spec or do some searching on-line about which of these two to use, you will be forgiven for choosing the Implicit Grant. It would appear to be the correct choice for single-page apps. However, if you do choose the Implicit Grant you'll be making my job easier. You see, the Implicit Grant passes the access token directly to the redirect_uri in the hash fragment. Security researchers have found many ways to get that raw value from the browser; using window.history is one such mechanism. This is called "access token leakage". The best current practice draft for OAuth 2 clients essentially states that you shouldn't ever use the Implicit Grant to fetch access tokens. Instead, you should use the only remaining choice - the "Authorization Code Grant".
+
+The Authorization Code Grant works by passing a one-time use "code" to the redirect_uri. The code is then used by the client to make a separate POST request to the AS token endpoint, which will finally return the access token as part of the response body. This means of returning the raw access token to the browser is much more secure than the Implicit Grant, since POST responses are never logged or stored in browser memory like the hash fragment.
+
+However, just using the authorization code grant is not enough. What happens if I get the code before your client uses it (maybe from your HTTP server logs)? Then I can submit it to the token endpoint first and get the access token. The solution to this attack is to use "Proof Key for Code Exchange" or PKCE. PKCE was originally designed to prevent attacks on mobile apps, but it applies just as well for single page apps. PKCE works by binding the token adding additional request details to the authorization request as well as to the token request. The end result is that only the original client which made the first request can complete the second; the code is insufficient.
```

Comment:
Are you just collecting thoughts for the moment, or is it a finished piece? If the latter, it doesn't seem to be referenced from the main README.

Reply:
Comments in regard to some other files not affected by this PR:

`service_workers.md`
"The act of adding a bearer token to a XHR request is more complicated than it seems at first." > "The act of adding a bearer token to an XHR request is more complicated than it seems at first." XHR stands for XMLHTTPRequest, and XML stands for eXtensible Markup Language; hence, however one pronounces the abbreviation, it starts with a vowel sound. 

`appAuthHelper.js`
Line #10: 
“* @param {Object} config - configation needed for working with the OP” > “* @param {Object} config - configuration needed for working with the OP”

Reply:
I also have a thought in regard to the origin isolation topic, but it may deserve a separate discussion/PR.